ember-core/shared/src/main/scala/org/http4s/ember/core/Parser.scala

```diff
@@ -62,26 +66,52 @@ private[ember] object Parser {
   )
 
   object HeaderP {
+    import scala.collection.mutable.ListBuffer
 
     private[this] val colon: Byte = ':'.toByte
     private[this] val contentLengthS = "Content-Length"
     private[this] val transferEncodingS = "Transfer-Encoding"
     private[this] val chunkedS = "chunked"
 
-    def parseHeaders[F[_]](message: Array[Byte], initIndex: Int, maxHeaderSize: Int)(implicit
+    final case class ParserState(
+        idx: Int,
+        state: Boolean,
+        throwable: Throwable,
+        complete: Boolean,
+        chunked: Boolean,
+        contentLength: Option[Long],
+        headers: ListBuffer[Header.Raw],
```

Comment:
This is a mutable variable. We do not have so many headers, this cant be passed in its immutable variant.

Reply:
Yeah I'd wondered about that. Will fix 👍

Reply:
Did you just mean `List` sorry? Currently the tests fail because the order of the headers is now reversed. Do we guarantee that we  preserve the request order of headers?

Reply:
If we agree that we don't guarantee anything about the order of headers, I could pick up https://github.com/http4s/http4s/pull/6201 again. :thinking:

Reply:
Glad you saw this @wjoel, I was exactly thinking about that PR as well :)

Reply:
Reverse when you complete. Shouldn't make a noticeable difference.